### PR TITLE
Implement json_pack_join and json_nest_join

### LINKED ISSUE #917: `json_pack_join()` and `json_nest_join()`
For #913

A wrapper around nest_join()

### PR DESCRIPTION
**Pull request by moodymudskipper**
Closes #917 

Contains :
* Function definitions (s3 generics + data.frame methods for consistency)
* Documentation
* Tests

### CODE REVIEW

**Review by krlmlr — COMMENTED** [review]
Thanks, let's do another round.

File: R/json-joins.R [hunk 1]
```diff
@@ -0,0 +1,32 @@
+#' JSON joins
+#'
+#' These are wrappers around `pack_join()` and `dplyr::nest_join()` which store
+#' the joined data into a json column.
+#' `json_pack_join()` returns all rows and columns in x with a new json columns that contains all packed matches from y.
+#' `json_nest_join()` returns all rows and columns in x with a new json columns that contains all nested matches from y.
+#'
+#' @inheritParams dplyr::nest_join
+#' @export
+json_nest_join <- function(x, y, by = NULL, copy = FALSE, keep = FALSE, name = NULL, ...) {
```
[the remark this answers is not in the document]
Comment (moodymudskipper) [reply R/json-joins.R:10]: We can but we lose consistency with `dplyr::nest_join()`

[the remark this answers is not in the document]
Comment (krlmlr) [reply R/json-joins.R:10]: `nest_join()` has it wrong in this case, I believe.

File: R/json-joins.R [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  UseMethod("json_nest_join")
+}
+
+#' @export
+json_nest_join.data.frame <- function(x, y, by = NULL, copy = FALSE, keep = FALSE, name = NULL, ...) {
```
[the remark this answers is not in the document]
Comment (moodymudskipper) [reply R/json-joins.R:15]: It's forwarded to `nest_join()` but `nest_join()` doesn't act on  it, which is a bug as far as I can tell:

``` r
library(dplyr, warn.conflicts = FALSE)
library(dm, warn.conflicts = FALSE)
df1 <- tibble(col1 = 1:2, key = letters[1:2])
df2 <- tibble(col2 = c(3:4, pi), col3 = c("X", "Y", "Z"), key2 = letters[c(1,1:2)])
nest_join(df1, df2, by = c(key = "key2"), keep = TRUE)
#> # A tibble: 2 × 3
#>    col1 key   df2             
#>   <int> <chr> <list>          
#> 1     1 a     <tibble [2 × 3]>
#> 2     2 b     <tibble [1 × 3]>

left_join(df1, df2, by = c(key = "key2"), keep = TRUE)
#> # A tibble: 3 × 5
#>    col1 key    col2 col3  key2 
#>   <int> <chr> <dbl> <chr> <chr>
#> 1     1 a      3    X     a    
#> 2     1 a      4    Y     a    
#> 3     2 b      3.14 Z     b

pack_join(df1, df2, by = c(key = "key2"), keep = TRUE)
#> # A tibble: 3 × 4
#>    col1 key   key2  df2$col2 $col3
#>   <int> <chr> <chr>    <dbl> <chr>
#> 1     1 a     a         3    X    
#> 2     1 a     a         4    Y    
#> 3     2 b     b         3.14 Z
```

<sup>Created on 2022-04-12 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

File: tests/testthat/_snaps/json-joins.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
+# json joins work
+
+    Code
+      packed <- json_pack_join(df1, df2, by = "key")
+      packed
+    Output
+      # A tibble: 3 x 3
+         col1 key   df2         
+        <int> <chr> <named list>
```
[the remark this answers is not in the document]
Comment (moodymudskipper) [reply tests/testthat/_snaps/json-joins.md]: split.data.frame() names its output, I unnamed it in https://github.com/cynkra/dm/pull/918/commits/4f888b12972d598224c4de3f10f7d2aa28b8ed10

File: R/json-joins.R [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+#' @rdname json_nest_join
+json_pack_join <- function(x, y, by = NULL, copy = FALSE, keep = FALSE, name = NULL, ...) {
+  UseMethod("json_pack_join")
+}
+
+#' @export
+json_pack_join.data.frame <- function(x, y, by = NULL, copy = FALSE, keep = FALSE, name = NULL, ...) {
+  name_var <- name %||% as_label(enexpr(y))
+  pack_join(x, y, by, copy, keep, name_var, ...) %>%
+    mutate(!!name_var := map(split(!!sym(name_var), 1:n()), jsonlite::toJSON))
```
[the remark this answers is not in the document]
Comment (moodymudskipper) [reply R/json-joins.R]: Implemented in https://github.com/cynkra/dm/pull/918/commits/72387f0c2d2c377cfcbaafdaa33e5761a82c7cf9

**Comment by moodymudskipper** [comment]
I oversaw that with `nest_join()` `keep = TRUE` means that the key column is kept inside of the **nested** data frames, so it's not a bug, barely slightly ambiguous documentation.

So to answer your question about what `keep` does for our `json_nest_join()` :  

```
df1 <- tibble(col1 = 1:2, key = letters[1:2])
df2 <- tibble(col2 = c(3:4, pi), col3 = c("X", "Y", "Z"), key = letters[c(1,1:2)])
waldo::compare(json_nest_join(df1, df2, by = "key"), json_nest_join(df1, df2, by = "key", keep = TRUE))
#> old$df2[[1]] vs new$df2[[1]]
#> - "[{\"col2\":3,\"col3\":\"X\"},{\"col2\":4,\"col3\":\"Y\"}]"
#> + "[{\"col2\":3,\"col3\":\"X\",\"key\":\"a\"},{\"col2\":4,\"col3\":\"Y\",\"key\":\"a\"}]"
#> 
#> `old$df2[[2]]`: "[{\"col2\":3.14159265358979,\"col3\":\"Z\"}]"              
#> `new$df2[[2]]`: "[{\"col2\":3.14159265358979,\"col3\":\"Z\",\"key\":\"b\"}]"
```

File: R/json-joins.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (moodymudskipper) [reply R/json-joins.R]: Answered in the general thread

File: R/json-joins.R [hunk 5]
```diff
@@ -0,0 +1,40 @@
+#' JSON joins
+#'
+#' These are wrappers around `pack_join()` and `dplyr::nest_join()` which store
+#' the joined data into a json column.
+#' `json_pack_join()` returns all rows and columns in x with a new json columns that contains all packed matches from y.
+#' `json_nest_join()` returns all rows and columns in x with a new json columns that contains all nested matches from y.
```
Comment (krlmlr) [inline R/json-joins.R:6]: For `dm_wrap_tbl()` and friends I opted for a separate help page for each verb, and also for separate files for each function, even if it becomes slightly redundant. It's easier in the long run.

File: R/json-joins.R [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+#'
+#' @inheritParams dplyr::nest_join
+#' @export
+json_nest_join <- function(x, y, by = NULL, ..., copy = FALSE, keep = FALSE, name = NULL) {
```
Comment (krlmlr) [inline R/json-joins.R:10]: Can you please add examples?

File: R/json-joins.R [hunk 7]
```diff
@@ -0,0 +1,40 @@
+#' JSON joins
+#'
+#' These are wrappers around `pack_join()` and `dplyr::nest_join()` which store
```
Comment (krlmlr) [inline R/json-joins.R:3]: "experimental" lifecycle

**Review by krlmlr — COMMENTED** [review]
Thanks, great! Just a bit of documentation work, looks ready to merge otherwise.

File: R/json-joins.R (hunk 5 above)
Comment (moodymudskipper) [reply R/json-joins.R:6]: done in https://github.com/cynkra/dm/pull/918/commits/4f51a4098013f885f8a619e05f5fd00673828c7a and https://github.com/cynkra/dm/pull/918/commits/16312a8094bc2f70713d201fe6d9a9dabdef96dc

File: R/json-joins.R (hunk 6 above)
Comment (moodymudskipper) [reply R/json-joins.R:10]: done in https://github.com/cynkra/dm/pull/918/commits/4f51a4098013f885f8a619e05f5fd00673828c7a

File: R/json-joins.R (hunk 7 above)
Comment (moodymudskipper) [reply R/json-joins.R:3]: done in https://github.com/cynkra/dm/pull/918/commits/4f51a4098013f885f8a619e05f5fd00673828c7a

File: R/json-joins.R [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+#' @export
+json_pack_join.data.frame <- function(x, y, by = NULL, ..., copy = FALSE, keep = FALSE, name = NULL) {
+  check_dots_empty()
+  name_var <- name %||% as_label(enexpr(y))
+  pack_join(x, y, by, copy, keep, name_var, ...) %>%
+    mutate(
+      !!name_var := map(
+        unname(split(!!sym(name_var), seq2(1, n()))),
```
[the remark this answers is not in the document]
Comment (moodymudskipper) [reply R/json-joins.R:35]: done in https://github.com/cynkra/dm/pull/918/commits/562c633b4430e4bdcaf65b96f1e4904f38a43bad

**Comment by moodymudskipper** [comment]
The first failure is the same as in #921 , due to vctrs version probably
The second failure is due to missing topics in pkgdown, presumably because I've updated the doc ? What is the correct workflow here?

```
-- Building function reference -------------------------------------------------
Error in check_missing_topics(rows, pkg) : 
  Topics missing from index: json_nest_join
• Topics missing from index: json_pack_join
```

**Comment by krlmlr** [comment]
Do you need to update `_pkgdown.yaml` ?

**Comment by krlmlr** [comment]
There is still a test failure in https://github.com/cynkra/dm/runs/6034481749?check_suite_focus=true#step:11:122.

I'm sorry: we also could use `json_nest()` and `json_pack()`, without the "join" part, because these operations can be useful for databases that don't support nested tables. Separate PR, or is it easy enough to do here?

**Comment by moodymudskipper** [comment]
Let's add it here

**Comment by moodymudskipper** [comment]
Should I implement `json_unest()` and `json_unpack()` in another PR ?

**Comment by krlmlr** [comment]
Yeah, this is less important for now.

The tests still fail, some error in the `.Rd` .

File: DESCRIPTION [hunk 9]
```diff
@@ -48,6 +48,7 @@ Imports:
     ellipsis,
     glue,
     igraph,
+    jsonlite,
```
Comment (krlmlr) [inline DESCRIPTION:51]: We're close to 20 imports, should we make this optional?

File: R/json_nest.R [hunk 10]
```diff
@@ -0,0 +1,32 @@
+# FIXME: once tbl_lazy method is implemented, update doc of `.data` arg
```
Comment (krlmlr) [inline R/json_nest.R:1]: No need to postpone this, implementing this is on the agenda :-)

File: R/json_nest_join.R [hunk 11]
```diff
@@ -0,0 +1,29 @@
+#' JSON nest join
+#'
+#' @description
+#' `r lifecycle::badge("experimental")`
+#'
+#' A wrapper around `dplyr::nest_join()` which stores the joined data into a JSON column.
```
Comment (krlmlr) [inline R/json_nest_join.R:6]: Here and elsewhere?

```suggestion
#' A wrapper around [dplyr::nest_join()] which stores the joined data into a JSON column.
```

File: pkgdown/_pkgdown.yml [hunk 12]
```diff
@@ -104,6 +104,14 @@ reference:
   - dm_unnest_tbl
   - dm_unpack_tbl
 
+- title: Data frames operations
```
Comment (krlmlr) [inline pkgdown/_pkgdown.yml:107]: Operations on data frames and lazy tables

File: tests/testthat/_snaps/json.md [hunk 13]
```diff
@@ -0,0 +1,88 @@
+# json joins work
+
+    Code
+      packed
+    Output
+      # A tibble: 3 x 3
+         col1 key   df2       
+        <int> <chr> <list>    
+      1     1 a     <json [1]>
```
Comment (krlmlr) [inline tests/testthat/_snaps/json.md:9]: Should we return a plain character here instead, so that a preview is shown when printed?

**Review by krlmlr — APPROVED** [review]
Thanks! Some high-level nitpicks, good to merge after that.

Please don't forget to add the NEWS entry to the merge commit message.

File: DESCRIPTION (hunk 9 above)
Comment (moodymudskipper) [reply DESCRIPTION]: Done in https://github.com/cynkra/dm/pull/918/commits/58ec7a5bb3f6bbbd98fbc7ed4128d927ffde8d81

File: R/json_nest.R (hunk 10 above)
Comment (moodymudskipper) [reply R/json_nest.R]: Done in https://github.com/cynkra/dm/pull/918/commits/80af53f925d7fbd3359d353dc957a4afb17487a1

File: R/json_nest_join.R (hunk 11 above)
Comment (moodymudskipper) [reply R/json_nest_join.R]: done in https://github.com/cynkra/dm/pull/918/commits/80af53f925d7fbd3359d353dc957a4afb17487a1

File: pkgdown/_pkgdown.yml (hunk 12 above)
Comment (moodymudskipper) [reply pkgdown/_pkgdown.yml]: done in https://github.com/cynkra/dm/pull/918/commits/d5a30be53c3975166e9c69e2480f532a314ef29f

File: tests/testthat/_snaps/json.md (hunk 13 above)
Comment (moodymudskipper) [reply tests/testthat/_snaps/json.md:9]: done in https://github.com/cynkra/dm/pull/918/commits/35d870958c1cc6f9fd00ca0d97c8f8406e63b944

File: R/json_nest.R [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+#' df <- tibble::tibble(x = c(1, 1, 1, 2, 2, 3), y = 1:6, z = 6:1)
+#' nested <- json_nest(df, data = c(y, z))
+#' nested
+json_nest <- function(.data, ..., .names_sep = NULL) {
+  UseMethod("json_nest")
+}
+
+#' @export
+json_nest.data.frame <- function(.data, ..., .names_sep = NULL) {
+  dot_nms <- ...names()
```
Comment (krlmlr) [inline R/json_nest.R:25]: `check_suggested()` here and elsewhere??

**Review by krlmlr — DISMISSED** [review]
Looks great, thanks!

File: R/json_nest.R (hunk 14 above)
Comment (moodymudskipper) [reply R/json_nest.R:26]: oops! done in https://github.com/cynkra/dm/pull/918/commits/e3c35260c34270cb02ebd06b7fe17bcf302a8b52

File: R/json_nest.R [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+#' `r lifecycle::badge("experimental")`
+#'
+#' A wrapper around [tidyr::nest()] which stores the nested data into JSON columns.
+#'
+#' @param .data A data frame, a data frame extension (e.g. a tibble), or  a lazy data frame (e.g. from dbplyr or dtplyr).
+#' @param .names_sep If `NULL`, the default, the names will be left as is.
+#' @param ... <[`tidy-select`][tidyr_tidy_select]> Columns to pack, specified
+#'   using name-variable pairs of the form `new_col = c(col1, col2, col3)`.
+#'   The right hand side can be any valid tidy select expression.
+#' @seealso [tidyr::nest], [json_nest_join]
```
Comment (krlmlr) [inline R/json_nest.R:13]: Tweaking later:

```suggestion
#' @seealso [tidyr::nest()], [json_nest_join()]
```

File: R/json_pack_join.R [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+#' df3 <- json_pack_join(df1, df2)
+#' df3
+#' df3$df2
+json_pack_join <- function(x, y, by = NULL, ..., copy = FALSE, keep = FALSE, name = NULL) {
+  UseMethod("json_pack_join")
+}
+
+#' @export
+json_pack_join.data.frame <- function(x, y, by = NULL, ..., copy = FALSE, keep = FALSE, name = NULL) {
+  check_suggested("jsonlite", TRUE)
```
Comment (krlmlr) [inline R/json_pack_join.R:25]: What does `TRUE` mean here? Logical literals passed as arguments should always be named.

Fixing later.

File: tests/testthat/test-json.R [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+
+  expect_snapshot({
+    df <- tibble::tibble(x1 = 1:3, x2 = 4:6, x3 = 7:9, y = 1:3)
+    packed <- json_pack(df, x = c(x1, x2, x3), y = y)
+    packed
+  })
+})
+
+test_that("`json_nest()` works", {
+  skip_if_not_installed("rlang", "0.99.0.9003")
```
Comment (krlmlr) [inline tests/testthat/test-json.R:37]: We're requiring rlang >= 1.0.1. Fixing later.

**Review by krlmlr — APPROVED** [review]
Thanks!

**Pull request merged**